File: sdk/apple/Sources/NotchAgentDeskProtocol/DeskFrameCodec.swift
import Foundation

public enum DeskProtocolContract {
    public static let product = "NotchAgent Desk"
    public static let major: UInt8 = 1
    public static let minor: UInt8 = 1
    public static let maximumPayloadBytes = 16 * 1_024
}

public enum DeskFrameType: UInt8, Sendable {
    case hello = 1
    case helloAcknowledgement = 2
    case snapshot = 3
    case heartbeat = 4
    case deviceTelemetry = 5
}

public struct DeskFrame: Sendable, Equatable {
    public var type: DeskFrameType
    public var sequence: UInt32
    public var payload: Data

    public init(type: DeskFrameType, sequence: UInt32, payload: Data) {
        self.type = type
        self.sequence = sequence
        self.payload = payload
    }
}

public enum DeskFrameCodecError: Error, Equatable {
    case invalidCOBS
    case invalidHeader
    case incompatibleProtocol
    case payloadTooLarge
    case invalidLength
    case invalidChecksum
}

public enum DeskFrameCodec {
    private static let magic: [UInt8] = [0x4E, 0x41, 0x44, 0x4B]
    private static let headerSize = 14

    public static func encode(_ frame: DeskFrame) throws -> Data {
        guard frame.payload.count <= DeskProtocolContract.maximumPayloadBytes else {
            throw DeskFrameCodecError.payloadTooLarge
        }
        var bytes = magic
        bytes.append(DeskProtocolContract.major)
        bytes.append(frame.type.rawValue)
        appendLittleEndian(frame.sequence, to: &bytes)
        appendLittleEndian(UInt32(frame.payload.count), to: &bytes)
        bytes.append(contentsOf: frame.payload)
        appendLittleEndian(crc32(bytes), to: &bytes)
        var encoded = cobsEncode(bytes)
        encoded.append(0)
        return Data(encoded)
    }

    public static func decodePacket(_ packet: Data) throws -> DeskFrame {
        let bytes = try cobsDecode(Array(packet))
        guard bytes.count >= headerSize + 4, Array(bytes.prefix(4)) == magic else {
            throw DeskFrameCodecError.invalidHeader
        }
        guard bytes[4] == DeskProtocolContract.major else {
            throw DeskFrameCodecError.incompatibleProtocol
        }
        guard let type = DeskFrameType(rawValue: bytes[5]) else {
            throw DeskFrameCodecError.invalidHeader
        }
        let sequence = readLittleEndian(bytes, at: 6)
        let payloadLength = Int(readLittleEndian(bytes, at: 10))
        guard payloadLength <= DeskProtocolContract.maximumPayloadBytes else {
            throw DeskFrameCodecError.payloadTooLarge
        }
        let expected = headerSize + payloadLength + 4
        guard bytes.count == expected else { throw DeskFrameCodecError.invalidLength }
        let checksum = readLittleEndian(bytes, at: expected - 4)
        guard checksum == crc32(Array(bytes.dropLast(4))) else {
            throw DeskFrameCodecError.invalidChecksum
        }
        return DeskFrame(
            type: type,
            sequence: sequence,
            payload: Data(bytes[headerSize..<(headerSize + payloadLength)])
        )
    }

    public static func cobsEncode(_ input: [UInt8]) -> [UInt8] {
        var output: [UInt8] = [0]
        var codeIndex = 0
        var code: UInt8 = 1
        for byte in input {
            if byte == 0 {
                output[codeIndex] = code
                codeIndex = output.count
                output.append(0)
                code = 1
            } else {
                output.append(byte)
                code &+= 1
                if code == 0xFF {
                    output[codeIndex] = code
                    codeIndex = output.count
                    output.append(0)
                    code = 1
                }
            }
        }
        output[codeIndex] = code
        return output
    }

    public static func cobsDecode(_ input: [UInt8]) throws -> [UInt8] {
        guard !input.isEmpty else { throw DeskFrameCodecError.invalidCOBS }
        var output: [UInt8] = []
        var index = 0
        while index < input.count {
            let code = Int(input[index])
            guard code != 0, index + code <= input.count + 1 else {
                throw DeskFrameCodecError.invalidCOBS
            }
            index += 1
            let end = index + code - 1
            guard end <= input.count else { throw DeskFrameCodecError.invalidCOBS }
            output.append(contentsOf: input[index..<end])
            index = end
            if code != 0xFF, index < input.count { output.append(0) }
        }
        return output
    }

    private static func appendLittleEndian(_ value: UInt32, to bytes: inout [UInt8]) {
        bytes.append(UInt8(truncatingIfNeeded: value))
        bytes.append(UInt8(truncatingIfNeeded: value >> 8))
        bytes.append(UInt8(truncatingIfNeeded: value >> 16))
        bytes.append(UInt8(truncatingIfNeeded: value >> 24))
    }

    private static func readLittleEndian(_ bytes: [UInt8], at offset: Int) -> UInt32 {
        UInt32(bytes[offset])
            | UInt32(bytes[offset + 1]) << 8
            | UInt32(bytes[offset + 2]) << 16
            | UInt32(bytes[offset + 3]) << 24
    }

    private static func crc32(_ bytes: [UInt8]) -> UInt32 {
        var crc: UInt32 = 0xFFFF_FFFF
        for byte in bytes {
            crc ^= UInt32(byte)
            for _ in 0..<8 {
                crc = (crc & 1) == 1 ? (crc >> 1) ^ 0xEDB8_8320 : crc >> 1
            }
        }
        return crc ^ 0xFFFF_FFFF
    }
}

public struct DeskFrameStreamDecoder: Sendable {
    private var buffer: [UInt8] = []

    public init() {}

    public mutating func append(_ data: Data) -> [Result<DeskFrame, DeskFrameCodecError>] {
        buffer.append(contentsOf: data)
        if buffer.count > (DeskProtocolContract.maximumPayloadBytes + 64) * 2 {
            buffer.removeAll(keepingCapacity: true)
            return [.failure(.payloadTooLarge)]
        }
        var results: [Result<DeskFrame, DeskFrameCodecError>] = []
        while let delimiter = buffer.firstIndex(of: 0) {
            let packet = Data(buffer[..<delimiter])
            buffer.removeFirst(delimiter + 1)
            guard !packet.isEmpty else { continue }
            do { results.append(.success(try DeskFrameCodec.decodePacket(packet))) }
            catch let error as DeskFrameCodecError { results.append(.failure(error)) }
            catch { results.append(.failure(.invalidHeader)) }
        }
        return results
    }
}

File: sdk/dotnet/NotchAgent.Desk.Protocol/DeskFrameCodec.cs
using System.Buffers.Binary;

namespace NotchAgent.Desk.Protocol;

public static class DeskProtocolContract
{
    public const string Product = "NotchAgent Desk";
    public const byte Major = 1;
    public const byte Minor = 1;
    public const int MaximumPayloadBytes = 16 * 1024;
}

public enum DeskFrameType : byte
{
    Hello = 1,
    HelloAcknowledgement = 2,
    Snapshot = 3,
    Heartbeat = 4,
    DeviceTelemetry = 5,
}

public sealed record DeskFrame(DeskFrameType Type, uint Sequence, byte[] Payload);

public sealed class DeskProtocolException(string message) : Exception(message);

public static class DeskFrameCodec
{
    private static readonly byte[] Magic = [0x4E, 0x41, 0x44, 0x4B];
    private const int HeaderSize = 14;

    public static byte[] Encode(DeskFrame frame)
    {
        ArgumentNullException.ThrowIfNull(frame);
        if (frame.Payload.Length > DeskProtocolContract.MaximumPayloadBytes)
            throw new DeskProtocolException("Payload exceeds the 16 KiB protocol limit.");

        var raw = new byte[HeaderSize + frame.Payload.Length + sizeof(uint)];
        Magic.CopyTo(raw, 0);
        raw[4] = DeskProtocolContract.Major;
        raw[5] = (byte)frame.Type;
        BinaryPrimitives.WriteUInt32LittleEndian(raw.AsSpan(6, 4), frame.Sequence);
        BinaryPrimitives.WriteUInt32LittleEndian(raw.AsSpan(10, 4), checked((uint)frame.Payload.Length));
        frame.Payload.CopyTo(raw, HeaderSize);
        BinaryPrimitives.WriteUInt32LittleEndian(raw.AsSpan(raw.Length - 4, 4), Crc32(raw.AsSpan(0, raw.Length - 4)));

        var encoded = CobsEncode(raw);
        Array.Resize(ref encoded, encoded.Length + 1);
        return encoded;
    }

    public static DeskFrame DecodePacket(ReadOnlySpan<byte> packet)
    {
        var raw = CobsDecode(packet);
        if (raw.Length < HeaderSize + 4 || !raw.AsSpan(0, 4).SequenceEqual(Magic))
            throw new DeskProtocolException("Invalid frame header.");
        if (raw[4] != DeskProtocolContract.Major)
            throw new DeskProtocolException("Incompatible protocol major.");
        if (!Enum.IsDefined(typeof(DeskFrameType), raw[5]))
            throw new DeskProtocolException("Unknown frame type.");

        var payloadLength = BinaryPrimitives.ReadUInt32LittleEndian(raw.AsSpan(10, 4));
        if (payloadLength > DeskProtocolContract.MaximumPayloadBytes)
            throw new DeskProtocolException("Payload exceeds the protocol limit.");
        if (raw.Length != HeaderSize + payloadLength + sizeof(uint))
            throw new DeskProtocolException("Invalid frame length.");

        var expected = BinaryPrimitives.ReadUInt32LittleEndian(raw.AsSpan(raw.Length - 4, 4));
        if (expected != Crc32(raw.AsSpan(0, raw.Length - 4)))
            throw new DeskProtocolException("Invalid CRC-32 checksum.");

        var payload = raw.AsSpan(HeaderSize, checked((int)payloadLength)).ToArray();
        return new DeskFrame(
            (DeskFrameType)raw[5],
            BinaryPrimitives.ReadUInt32LittleEndian(raw.AsSpan(6, 4)),
            payload
        );
    }

    public static byte[] CobsEncode(ReadOnlySpan<byte> input)
    {
        var output = new List<byte>(input.Length + input.Length / 254 + 1) { 0 };
        var codeIndex = 0;
        byte code = 1;
        foreach (var value in input)
        {
            if (value == 0)
            {
                output[codeIndex] = code;
                codeIndex = output.Count;
                output.Add(0);
                code = 1;
            }
            else
            {
                output.Add(value);
                code++;
                if (code == 0xFF)
                {
                    output[codeIndex] = code;
                    codeIndex = output.Count;
                    output.Add(0);
                    code = 1;
                }
            }
        }
        output[codeIndex] = code;
        return [.. output];
    }

    public static byte[] CobsDecode(ReadOnlySpan<byte> input)
    {
        if (input.IsEmpty) throw new DeskProtocolException("Empty COBS packet.");
        var output = new List<byte>(input.Length);
        var index = 0;
        while (index < input.Length)
        {
            var code = input[index];
            if (code == 0 || index + code > input.Length + 1)
                throw new DeskProtocolException("Invalid COBS packet.");
            index++;
            var end = index + code - 1;
            if (end > input.Length) throw new DeskProtocolException("Invalid COBS packet.");
            while (index < end) output.Add(input[index++]);
            if (code != 0xFF && index < input.Length) output.Add(0);
        }
        return [.. output];
    }

    private static uint Crc32(ReadOnlySpan<byte> bytes)
    {
        var crc = 0xFFFF_FFFFu;
        foreach (var value in bytes)
        {
            crc ^= value;
            for (var bit = 0; bit < 8; bit++)
                crc = (crc & 1) == 1 ? (crc >> 1) ^ 0xEDB8_8320u : crc >> 1;
        }
        return crc ^ 0xFFFF_FFFFu;
    }
}

public sealed class DeskFrameStreamDecoder
{
    private readonly List<byte> _buffer = [];

    public IReadOnlyList<DeskFrame> Append(ReadOnlySpan<byte> bytes)
    {
        _buffer.AddRange(bytes.ToArray());
        if (_buffer.Count > (DeskProtocolContract.MaximumPayloadBytes + 64) * 2)
        {
            _buffer.Clear();
            throw new DeskProtocolException("Frame stream exceeded its bounded buffer.");
        }

        var frames = new List<DeskFrame>();
        while (_buffer.IndexOf(0) is var delimiter && delimiter >= 0)
        {
            var packet = _buffer.GetRange(0, delimiter).ToArray();
            _buffer.RemoveRange(0, delimiter + 1);
            if (packet.Length > 0) frames.Add(DeskFrameCodec.DecodePacket(packet));
        }
        return frames;
    }
}
